### PR TITLE
Added QuicException.TransportErrorCode

### LINKED ISSUE #87262: [API Proposal]: Add transport error code to QuicException
### Background and motivation

Related to #72666. Quic RFC (https://www.rfc-editor.org/rfc/rfc9000.html#name-transport-error-codes) defines set of error codes and we probably should surface them in cases when `QuicConnection` fails for one of them.

We already have `ProtocolError` to cover some cases when peer sends invalid data and it feels like there are also cases when the transport fail for other reasons. The proposal is to remove  specific `ProtocolError` as Quic is still in `Preview` and replace it it with more generic `TransportError`. 

We already have `ApplicationErorrCode` in `QuicException` so the proposal is to add also `TransportErrorCode` 

### API Proposal

```diff
namespace System.Net.Quic
{
    public enum QuicError
    {
        // existing members omitted

-       /// <summary>
-       /// A QUIC protocol error was encountered.
-       /// </summary>
-       ProtocolError,
+       /// <summary>
+       /// Operation failed because peer transport error occurred.  
+       /// </summary>
+       TransportError,
  }
  public class QuicException
  {
        public System.Net.Quic.QuicError QuicError { get { throw null; } }
        public long? ApplicationErrorCode { get { throw null; } }
+.      public long? TransportErrorCode { get { throw null; }}
  }
}
```

### API Usage

```c#
  try {
    QuicConnection connection = await QuicConnection.ConnectAsync();
  }
  catch (QuicExeption ex)
  {
        log($"{ex.Message} Application error {ex.ApplicationErrorCode} transport error {ex.TransportErrorCode}");
  }
```

the transport errors are generally unrecoverable e.g. we do not expect separate code branches based on specific value. 

### Alternative Designs

```diff
namespace System.Net.Quic
{

  public class QuicException
  {
         public System.Net.Quic.QuicError QuicError { get { throw null; } }
-        public long? ApplicationErrorCode { get { throw null; } }
+.       public long? ErrorCode { get { throw null; }}
  }
}
```

We can unify the error code and the value would have meaning based on `QuicError` value. We generally felt this may be confusing and easy to make mistake and misinterpret the value. 

Since the error codes are applicable only ins some cases it may not make sense to drag nullable fields in al instances. 
We can subtype 

```diff
namespace System.Net.Quic
{

  public class QuicTransportException: QuicException
  {
         public System.Net.Quic.QuicError QuicError { get { throw null; } }
+.       public long TransportErrorCode { get { throw null; }}
  }
}

``` 

 We also talk about making the `TransportErrorCode` strongly typed e.g. creating `enum` for it. But it is not clear how stable `Quic` protocol is and we would need to keep up. Also RFC defines some ranges and that may be hard to map to `enum`. Current plan is to surface the reason in some more friendly way in text when we know it.
  

### Risks

_No response_

### PR DESCRIPTION
**Pull request by AlexRadch**
Resolves #87262 #72666

### CODE REVIEW

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs [hunk 1]
```diff
@@ -63,7 +63,9 @@ public enum QuicError
     public sealed partial class QuicException : System.IO.IOException
     {
         public QuicException(System.Net.Quic.QuicError error, long? applicationErrorCode, string message) { }
+        public QuicException(System.Net.Quic.QuicError error, long? applicationErrorCode, long? transportErrorCode, string message) { }
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:66]: This needs to be internal as #87262 did not approve new public constructor.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs [hunk 2]
```diff
@@ -38,21 +38,21 @@ internal static bool TryGetStreamExceptionForMsQuicStatus(int status, [NotNullWh
...
-            exception = ThrowHelper.GetExceptionForMsQuicStatus(status);
+            exception = GetExceptionForMsQuicStatus(status);
             return true;
         }
         exception = null;
         return false;
     }
 
-    internal static Exception GetExceptionForMsQuicStatus(int status, string? message = null)
+    internal static Exception GetExceptionForMsQuicStatus(int status, ulong errorCode = 0, string? message = null)
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs:53]: I'm wondering if this should be nullable in case there is error 0 ?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs [hunk 3]
```diff
@@ -494,7 +494,7 @@ private unsafe int HandleEventShutdownInitiatedByTransport(ref SHUTDOWN_INITIATE
     {
         // TODO: we should propagate transport error code.
         // https://github.com/dotnet/runtime/issues/72666
-        Exception exception = ExceptionDispatchInfo.SetCurrentStackTrace(ThrowHelper.GetExceptionForMsQuicStatus(data.Status));
+        Exception exception = ExceptionDispatchInfo.SetCurrentStackTrace(ThrowHelper.GetExceptionForMsQuicStatus(data.Status, data.ErrorCode));
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:497]: we should remove the comments above

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs [hunk 4]
```diff
@@ -31,18 +31,18 @@ namespace System.Net.Quic;
 /// </summary>
 /// <remarks>
 /// <see cref="QuicConnection" /> can either be accepted from <see cref="QuicListener.AcceptConnectionAsync(CancellationToken)" /> (inbound connection),
-/// or create with a static method <see cref="QuicConnection.ConnectAsync(System.Net.Quic.QuicClientConnectionOptions, CancellationToken)" /> (outbound connection).
+/// or create with a static method <see cref="ConnectAsync(QuicClientConnectionOptions, CancellationToken)" /> (outbound connection).
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:34]: why? I think this is needed so generated documentation is linked properly. 
@ManickaP  may know for sure....

**Review by wfurt — COMMENTED** [review]
please also look at https://github.com/dotnet/runtime/pull/87679#issuecomment-1596882837 and try to address it.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs (hunk 4 above)
Comment (AlexRadch) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:34]: The compiler suggests these changes.

**Comment by AlexRadch** [comment]
> please also look at [#87679 (comment)](https://github.com/dotnet/runtime/pull/87679#issuecomment-1596882837) and try to address it.

Thank you, I added the transport error code propagation.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs [hunk 5]
```diff
@@ -31,7 +31,7 @@ namespace System.Net.Quic;
 /// Apart from stream API, <see cref="QuicStream"/> also exposes QUIC specific features:
 /// <list type="bullet">
 /// <item>
-/// <term><see cref="WriteAsync(System.ReadOnlyMemory{byte},bool,System.Threading.CancellationToken)"/></term>
+/// <term><see cref="WriteAsync(ReadOnlyMemory{byte},bool,CancellationToken)"/></term>
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:34]: Please, revert these unrelated changes in triple slash comments.

**Review by ManickaP — COMMENTED** [review]
Looks good modulo comments, thanks!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs [hunk 6]
```diff
@@ -61,17 +61,17 @@ internal static Exception GetExceptionForMsQuicStatus(int status, string? messag
...
-            if (status == QUIC_STATUS_ALPN_IN_USE) return new QuicException(QuicError.AlpnInUse, null, SR.net_quic_protocol_error);
+            if (status == QUIC_STATUS_CONNECTION_REFUSED) return new QuicException(QuicError.ConnectionRefused, null, (long?)errorCode, SR.net_quic_connection_refused);
+            if (status == QUIC_STATUS_CONNECTION_TIMEOUT) return new QuicException(QuicError.ConnectionTimeout, null, (long?)errorCode, SR.net_quic_timeout);
+            if (status == QUIC_STATUS_VER_NEG_ERROR) return new QuicException(QuicError.VersionNegotiationError, null, (long?)errorCode, SR.net_quic_ver_neg_error);
+            if (status == QUIC_STATUS_CONNECTION_IDLE) return new QuicException(QuicError.ConnectionIdle, null, (long?)errorCode, SR.net_quic_connection_idle);
+            if (status == QUIC_STATUS_PROTOCOL_ERROR) return new QuicException(QuicError.TransportError, null, (long?)errorCode, SR.net_quic_protocol_error);
+            if (status == QUIC_STATUS_ALPN_IN_USE) return new QuicException(QuicError.AlpnInUse, null, (long?)errorCode, SR.net_quic_protocol_error);
```
Comment (rzikm) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs:74]: are the casts necessary here?

```suggestion
            if (status == QUIC_STATUS_CONNECTION_REFUSED) return new QuicException(QuicError.ConnectionRefused, null, errorCode, SR.net_quic_connection_refused);
            if (status == QUIC_STATUS_CONNECTION_TIMEOUT) return new QuicException(QuicError.ConnectionTimeout, null, errorCode, SR.net_quic_timeout);
            if (status == QUIC_STATUS_VER_NEG_ERROR) return new QuicException(QuicError.VersionNegotiationError, null, errorCode, SR.net_quic_ver_neg_error);
            if (status == QUIC_STATUS_CONNECTION_IDLE) return new QuicException(QuicError.ConnectionIdle, null, errorCode, SR.net_quic_connection_idle);
            if (status == QUIC_STATUS_PROTOCOL_ERROR) return new QuicException(QuicError.TransportError, null, errorCode, SR.net_quic_protocol_error);
            if (status == QUIC_STATUS_ALPN_IN_USE) return new QuicException(QuicError.AlpnInUse, null, errorCode, SR.net_quic_protocol_error);
```

**Review by rzikm — COMMENTED** [review]
LGTM modulo comments.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs [hunk 7]
```diff
@@ -61,17 +61,17 @@ internal static Exception GetExceptionForMsQuicStatus(int status, string? messag
 
         return ex;
 
-        static Exception GetExceptionInternal(int status, string? message)
+        static Exception GetExceptionInternal(int status, ulong? errorCode, string? message)
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs:64]:
```suggestion
        static Exception GetExceptionInternal(int status, long? errorCode, string? message)
```
Than the casts won't be necessary.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs [hunk 8]
```diff
@@ -38,21 +38,21 @@ internal static bool TryGetStreamExceptionForMsQuicStatus(int status, [NotNullWh
...
-            exception = ThrowHelper.GetExceptionForMsQuicStatus(status);
+            exception = GetExceptionForMsQuicStatus(status);
             return true;
         }
         exception = null;
         return false;
     }
 
-    internal static Exception GetExceptionForMsQuicStatus(int status, string? message = null)
+    internal static Exception GetExceptionForMsQuicStatus(int status, ulong? errorCode = default, string? message = null)
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs:53]:
```suggestion
    internal static Exception GetExceptionForMsQuicStatus(int status, long? errorCode = default, string? message = null)
```

**Review by ManickaP — COMMENTED** [review]
One last comment, otherwise looks good and we can merge this, thanks!

**Comment by karelz** [comment]
@AlexRadch will you be able to finish the changes? It is highly desirable to merge it by tomorrow 7/17 to catch our Platform complete deadline (= finish all features) for 8.0. Thanks!

**Comment by ManickaP** [comment]
I'm taking over, @wfurt has a pending PR on this, we need to get this in today.

**Review by ManickaP — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs [hunk 9]
```diff
@@ -67,11 +67,7 @@ internal static unsafe T GetMsQuicParameter<T>(MsQuicSafeHandle handle, uint par
             &length,
             (byte*)&value);
 
-        if (StatusFailed(status))
-        {
-            throw ThrowHelper.GetExceptionForMsQuicStatus(status, $"GetParam({handle}, {parameter}) failed");
-        }
-
+        ThrowHelper.ThrowIfMsQuicError(status, $"GetParam({handle}, {parameter}) failed");
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs:70]: This is now going to allocate a string for the error message even in the success case.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs [hunk 10]
```diff
@@ -84,9 +80,6 @@ internal static unsafe void SetMsQuicParameter<T>(MsQuicSafeHandle handle, uint
             (uint)sizeof(T),
             (byte*)&value);
 
-        if (StatusFailed(status))
-        {
-            throw ThrowHelper.GetExceptionForMsQuicStatus(status, $"SetParam({handle}, {parameter}) failed");
-        }
+        ThrowHelper.ThrowIfMsQuicError(status, $"SetParam({handle}, {parameter}) failed");
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs:83]: This is now going to allocate a string for the error message even in the success case.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs (hunk 9 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs:70]: I'm going to address this in some of my future PRs. Or @wfurt do you think you could revert these two in #88614?

Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs:70]: I think it is ok in interim as it is only perf optimization. I feel it is more important to get the API out.
